`RichTooltip/Popover` component - `PopoverPrimitive` component documentation [03b] (#2046)
hashicorp/design-system: website/docs/utilities/popover-primitive/partials/code/component-api.md

## Component API

<Doc::ComponentApi as |C|>
  <C.Property @name="isOpen" @type="boolean" @default="false" @values={{array "true" "false"}}>
    Controls if the popover should be rendered initially opened.
    <br />
    <br />

>     _Notice: in this case the popover can't be dismissed via `esc` or `click outside` until the end user has interacted with it (it's in a "manual" state)._

Reviewer: ```suggestion
    _Notice: in this case, the popover can't be dismissed via `esc` or `click outside` until the end user has interacted with it (it's in a "manual" state)._
```

  </C.Property>
  <C.Property @name="enableSoftEvents" @type="boolean" @default="false" @values={{array "true" "false"}}>

>     Assigns "soft" events listeners (`mouseEnter/Leave` + `focusIn/Out`) to the toggle, to control the visibility of the popover content.

Reviewer 2: Question: where does the term "soft event" come from?

Author: We came up with this term to refer to non-click events. Do you have suggestions for an alternative term?

Reviewer 2: I'm thinking about it now, I won't object if I can't figure out something else to propose. Right now it feels technically inaccurate so I will try.

Reviewer 2: What do you think about `enableIntentEvents`?  Intent is a more technically accurate term for what we're going for, WDYT?

Reviewer: I think "soft" makes decent sense since clicking is a more active "hard" interaction so hovering & focusing are less active so "softer"...

>     Assigns "soft" events listeners (`mouseEnter/Leave` + `focusIn/Out`) to the toggle, to control the visibility of the popover content.

Reviewer: ```suggestion
    Assigns "soft" event listeners (`mouseEnter/Leave` + `focusIn/Out`) to the toggle, to control the visibility of the popover content.
```

  </C.Property>
  <C.Property @name="enableClickEvents" @type="boolean" @default="false" @values={{array "true" "false"}}>

>     Assigns a "click" event listener (`onClick`) to the toggle, to control the visibility of the popover content.

Reviewer 2: I don't think we need the quotes here. 

```suggestion
    Assigns a click event listener (`onClick`) to the toggle, to control the visibility of the popover content.
```

Reviewer: Agree. I think there's an overuse of quotes overall in some of the text here.

  </C.Property>
  <C.Property @name="onOpen" @type="function">
    A callback function invoked when the popover is opened (if provided).
  </C.Property>
  <C.Property @name="onClose" @type="function">
    Provides a callback function invoked when the popover is closed (if provided).
  </C.Property>
  <C.Property @name="[PP].setupPrimitiveContainer" @type="modifier">

>     Provides a modifiers that needs to be applied to the container of the toggle and popover elements.

Reviewer: ```suggestion
    Provides a modifier that needs to be applied to the container of the toggle and popover elements.
```

  </C.Property>
  <C.Property @name="[PP].setupPrimitiveToggle" @type="modifier">

>     Provides a modifiers that needs to be applied to the toggle element.

Reviewer: ```suggestion
    Provides a modifier that needs to be applied to the toggle element.
```

    <br />
    <br />

>     _⚠️ Important: the HTML element must be a `<button>` for accessibility conformance. If not, the component will throw an errow._

Reviewer: ```suggestion
    _⚠️ Important: the HTML element must be a `<button>` for accessibility conformance. If not, the component will throw an error._
```

  </C.Property>
  <C.Property @name="[PP].setupPrimitivePopover" @type="modifier">

>     Provides a modifiers that needs to be applied to the popover element. It accepts an `anchoredPositionOptions` object as "named" argument, with different keys corresponding to different options (see below). These options are forwarded to underlying modifier `hds-anchored-position`, and in turn are passed down to the [Floating UI](https://floating-ui.com) library.

Reviewer: ```suggestion
    Provides a modifier that needs to be applied to the popover element. It accepts an `anchoredPositionOptions` object as "named" argument, with different keys corresponding to different options (see below). These options are forwarded to the underlying `hds-anchored-position` modifier, and in turn are passed down to the [Floating UI](https://floating-ui.com) library.
```

    <Doc::ComponentApi as |C|>
      <C.Property @name="anchoredPositionOptions.placement" @type="enum" @values={{array "top" "top-start" "top-end" "right" "right-start" "right-end" "bottom" "bottom-start" "bottom-end" "left" "left-start" "left-end"}} @default="bottom">
        Placement for the preferred starting position of the popover relative to the toggle element.
        <br />
        <br />
        _Notice: if `@enableCollision` is set, the popover will automatically shift position to remain visible when near the edges of the screen regardless of the starting placement._
      </C.Property>
      <C.Property @name="anchoredPositionOptions.strategy" @type="enum" @values={{array "absolute" "fixed"}} @default="absolute">

>         Controls how the layout of the popover is applied using the CSS `position` property. Since the component uses the [native web Popover API](https://developer.mozilla.org/en-US/docs/Web/API/Popover_API) which promotes the popover to the [top layer](https://developer.mozilla.org/en-US/docs/Glossary/Top_layer), there is no need to use the `fixed` position to avoid stacking context conflicts, but we leave the option open for edge cases in which this is needed anyway.

Reviewer: ```suggestion
        Controls how the layout of the popover is applied using the CSS `position` property. Since the component uses the [native web Popover API](https://developer.mozilla.org/en-US/docs/Web/API/Popover_API) which promotes the popover to the [top layer](https://developer.mozilla.org/en-US/docs/Glossary/Top_layer), there is no need to use the `fixed` position to avoid stacking context conflicts, but we leave the option open for edge cases in which this is needed.
```

        <br />
        <br />

>         _Notice: if the position/strategy is set to `fixed`, the rendering of the popover becomes jagged when the page is scrolled._

Reviewer: What does "jagged" mean in this context?

Author: that jumps when scrolling; what is the right term in English?

Reviewer 2: `janky` is the term used most often for this context.

Author: OK, I'll use that instead 👍

      </C.Property>
      <C.Property @name="anchoredPositionOptions.offsetOptions" @type="number|object" @default="0">

>         A number that represents the “distance” between the popover element and the toggle element. An object can also be passed, which enables to individually configure different axis.

Reviewer: ```suggestion
        A number that represents the distance between the popover element and the toggle element. An object can also be passed, which enables to individually configure different axis.
```

        <br />
        <br />
        For details see: [Floating UI > Offset > Options](https://floating-ui.com/docs/offset#options).
        <br />
        <br />
        _Notice: this options can be used to control the relative position of the arrow in relation to the toggle._
      </C.Property>
      <C.Property @name="anchoredPositionOptions.enableCollisionDetection" @type="boolean|string" @values={{array "true" "false" "flip" "shift" "auto"}} @default="false">

>         This property controls if the popover should automatically adapt its position to remain visible when near the edges of the screen. It can be enabled for both axes setting it to `true`, or a single axes can be chosen passing the `flip` or `shift` value. If set to `auto` it will automatically place the popover in the position where there's more space available, but in this case it will ignore the `placement` value.

Reviewer: ```suggestion
        This property controls if the popover should automatically adapt its position to remain visible when near the edges of the screen. It can be enabled for both axes by setting it to `true`, or a single axes can be chosen by passing either the `flip` or `shift` values. If set to `auto`, it will automatically place the popover in the position where there's more space available, but in this case it will ignore the `placement` value.
```

      </C.Property>
      <C.Property @name="anchoredPositionOptions.flipOptions" @type="object" @default="{ padding: 8 }">
        The options for the `flip` middleware in Floating UI, that controls the automatic repositioning of the popover along its side axis.
        <br />
        <br />
        For details about how this middleware works (and its options) see: [Floating UI > Flip](https://floating-ui.com/docs/flip)
      </C.Property>
      <C.Property @name="anchoredPositionOptions.shiftOptions" @type="object" @default="{ padding: 8, limiter: limitShift() }">
        The options for the `shift` middleware in Floating UI, that controls the automatic repositioning of the popover along its axis of alignment.
        <br />
        <br />
        For details about how this middleware works (and its options) see: [Floating UI > Shift](https://floating-ui.com/docs/shift)
      </C.Property>
      <C.Property @name="anchoredPositionOptions.autoPlacementOptions" @type="object" @default="{ padding: 8 }">

>         The options for the `autoPlacement` middleware in Floating UI, that controls the automatic repositioning of the popover based on the most space available.

Reviewer: ```suggestion
        The options for the `autoPlacement` middleware in Floating UI that controls the automatic repositioning of the popover based on the most space available.
```

        <br />
        <br />
        For details about how this middleware works (and its options) see: [Floating UI > Shift](https://floating-ui.com/docs/autoPlacement)
      </C.Property>
      <C.Property @name="anchoredPositionOptions.middlewareExtra" @type="array">
        An array of "middleware" functions to be executed by the Floating UI library.
        <br />
        <br />
        For details about how these functions work in the context of the library (and how to define custom ones) see: [Floating UI > Middleware](https://floating-ui.com/docs/middleware) and [Floating UI > Custom middleware](https://floating-ui.com/docs/computePosition#custom).
      </C.Property>
      <C.Property @name="anchoredPositionOptions.arrowElement" @type="DOM element">
        A reference to the DOM element that represents the "arrow" decoration, if it exists, so it position can be updated accordingly to the popover position in relation to the toggle anchor.
        <br />
        <br />
        For details see: [Floating UI > Arrow](https://floating-ui.com/docs/arrow)
      </C.Property>
      <C.Property @name="anchoredPositionOptions.arrowElement" @type="string">

>         A DOM selector to the "arrow", if it's not possible to provide directly a reference to the DOM element (internally it's then converted to the `arrowElement` option).

Reviewer: ```suggestion
        A DOM selector for the "arrow", if it's not possible to provide a direct reference to the DOM element (internally it's then converted to the `arrowElement` option).
```

      </C.Property>
      <C.Property @name="anchoredPositionOptions.arrowPadding" @type="number">
        The padding between the arrow and the edges of the popover element.
        <br />
        <br />
        For details see: [Floating UI > Arrow > Padding](https://floating-ui.com/docs/arrow#padding)
      </C.Property>
    </Doc::ComponentApi>
  </C.Property>
  <C.Property @name="[PP].toggleElement" @type="DOM element">
    Provides a reference to the toggle DOM element element (container).
  </C.Property>
  <C.Property @name="[PP].arrowElement" @type="DOM element">

>     Provides a reference to the arrow DOM element element (if provided to the `setupPrimitivePopover` modifier through the `anchoredPositionOptions` object).

Reviewer: ```suggestion
    Provides a reference to the arrow DOM element (if provided to the `setupPrimitivePopover` modifier through the `anchoredPositionOptions` object).
```

  </C.Property>
  <C.Property @name="[PP].popoverElement" @type="DOM element">

>     A reference to the popover DOM element element (container).

Reviewer: ```suggestion
    A reference to the popover DOM element (container).
```
